### PR TITLE
Adding Kinyarwanda TN, including cardinals, time, abbreviations and t…

### PR DESCRIPTION
**Pull request by kurt0cougar**
…ransliteration semiotic classes

# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Have you signed your commits? Use ``git commit -s`` to sign.
- [x] Do all unittests finish successfully before sending PR?
   1) ``pytest`` or (if your machine does not have GPU) ``pytest --cpu`` from the root folder (given you marked your test cases accordingly `@pytest.mark.run_only_on('CPU')`).
   2) Sparrowhawk tests ``bash tools/text_processing_deployment/export_grammars.sh --MODE=test ...``
- [x] If you are adding a new feature: Have you added test cases for both `pytest` and Sparrowhawk [here](tests/nemo_text_processing).
- [x] Have you added ``__init__.py`` for every folder and subfolder, including `data` folder which has .TSV files?
- [x] Have you followed codeQL results and removed unused variables and imports (report is at the bottom of the PR in github review box) ?
- [x] Have you added the correct license header `Copyright (c) 2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.` to all newly added Python files?
- [x] If you copied [nemo_text_processing/text_normalization/en/graph_utils.py](nemo_text_processing/text_normalization/en/graph_utils.py) your header's second line should be `Copyright 2015 and onwards Google, Inc.`. See an example [here](https://github.com/NVIDIA/NeMo-text-processing/blob/main/nemo_text_processing/text_normalization/en/graph_utils.py#L2).
- [x] Remove import guards (`try import: ... except: ...`) if not already done.
- [ ] If you added a new language or a new feature please update the [NeMo documentation](https://github.com/NVIDIA/NeMo/blob/main/docs/source/nlp/text_normalization/wfst/wfst_text_normalization.rst) (lives in different repo).
- [x] Have you added your language support to [tools/text_processing_deployment/pynini_export.py](tools/text_processing_deployment/pynini_export.py).
  

**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation
- [ ] Test

If you haven't finished some of the above items you can still open "Draft" PR.

### CODE REVIEW

**Comment by tbartley94** [comment]
@kurt0cougar Implementation question: what's the rational for keeping the cardinals as a separate file. Does reading from file improve graph construction speed that much?

**Comment by tbartley94** [comment]
@elenaltz Do we have a format for shared copyright headers?

**Comment by ekmb** [comment]
> @elenaltz Do we have a format for shared copyright headers?

@tbartley94 do you mean something like this https://github.com/NVIDIA/NeMo-text-processing/blob/main/nemo_text_processing/hybrid/mlm_scorer.py

File: nemo_text_processing/text_normalization/rw/verbalizers/time.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+from nemo_text_processing.text_normalization.en.graph_utils import (
+    GraphFst,
+    delete_space,
+    NEMO_CHAR
+)
+
+class VerbalizeTimeFst(GraphFst):
+    def __init__(self):
+        super().__init__(name="time",kind="verbalize")
+        hour = (pynutil.delete("hours:")+delete_space+pynutil.delete("\"")+pynini.closure(NEMO_CHAR,1,60)+pynutil.delete("\"")+delete_space+pynutil.delete("minutes:")+delete_space+pynutil.delete("\"")+pynini.closure(NEMO_CHAR,1,60)+pynutil.delete("\""))
```
Comment (tbartley94) [inline nemo_text_processing/text_normalization/rw/verbalizers/time.py:25]: can you break this up over several lines? difficult to read

File: tests/nemo_text_processing/rw/test_sparrowhawk_normalization.sh [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+
+testTNWhitelist() {
+  input=$TEST_DIR/data_text_normalization/test_cases_whitelist.txt
+  runtest $input
+}
+
+
+
+
```
Comment (tbartley94) [inline tests/nemo_text_processing/rw/test_sparrowhawk_normalization.sh:49]: Create a test for `TNWord` (helps catch overproductivity)

**Review by tbartley94 — CHANGES_REQUESTED** [review]
Added comments. Please update all headers to 2024 and include copyright mention for Digital Umuganda (both so y'all are properly accredited and so the MIT license is valid for OSS). Also please add tests for Words to make sure the model is properly processing non-normalizable texts.

**Comment by tbartley94** [comment]
@kurt0cougar Any questions on comments?

File: tests/nemo_text_processing/rw/data_text_normalization/test_cases_cardinal.txt [hunk 3]
```diff
@@ -0,0 +1,9 @@
+1~rimwe
```
[the remark this answers is not in the document]
Comment (kurt0cougar) [reply tests/nemo_text_processing/rw/data_text_normalization/test_cases_cardinal.txt:1]: I will include them, thanks

File: tools/text_processing_deployment/export_grammars.sh [hunk 4]
```diff
@@ -30,9 +30,9 @@
 # To test ITN grammars, run:
 #       bash export_grammars.sh --GRAMMARS=itn_grammars --LANGUAGE=en --MODE=test
 
-GRAMMARS="itn_grammars" # tn_grammars
-INPUT_CASE="lower_cased" # cased
-LANGUAGE="en" # language, {'en', 'es', 'de','zh'} supports both TN and ITN, {'pt', 'ru', 'fr', 'vi', 'mr'} supports ITN only
+GRAMMARS="tn_grammars" # tn_grammars
```
[the remark this answers is not in the document]
Comment (kurt0cougar) [reply tools/text_processing_deployment/export_grammars.sh:33]: Noted.

File: nemo_text_processing/text_normalization/rw/taggers/cardinal.py [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+                            "16": "icumi na gatandatu",
+                            "17": "icumi na karindwi",
+                            "18": "icumi na munani",
+                            "19": "icumi na icyenda",
+                        }
+            # Tens mapping
+            tens_to_word = {
+                "0": "",
+                "10": "icumi",
+                "20": "makumyabiri",
```
[the remark this answers is not in the document]
Comment (kurt0cougar) [reply nemo_text_processing/text_normalization/rw/taggers/cardinal.py:56]: No, it is never used. But I believe it is used in Kirundi.

File: tests/nemo_text_processing/rw/test_cardinal.py [hunk 6]
```diff
@@ -1,4 +1,4 @@
-# Copyright (c) 2021, NVIDIA CORPORATION.  All rights reserved.
+# Copyright (c) 2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
Comment (tbartley94) [inline tests/nemo_text_processing/rw/test_cardinal.py:1]: 2024 and Add Copyrights for Digital Umuganda (or parent org..if applicable)

File: tests/nemo_text_processing/rw/test_time.py [hunk 7]
```diff
@@ -1,4 +1,4 @@
-# Copyright (c) 2021, NVIDIA CORPORATION.  All rights reserved.
+# Copyright (c) 2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
Comment (tbartley94) [inline tests/nemo_text_processing/rw/test_time.py:1]: 2024 and Add Copyrights for Digital Umuganda (or parent org..if applicable)

File: tests/nemo_text_processing/rw/test_whitelist.py [hunk 8]
```diff
@@ -1,4 +1,4 @@
-# Copyright (c) 2021, NVIDIA CORPORATION.  All rights reserved.
+# Copyright (c) 2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
Comment (tbartley94) [inline tests/nemo_text_processing/rw/test_whitelist.py:1]: 2024 and Add Copyrights for Digital Umuganda (or parent org..if applicable)

File: nemo_text_processing/text_normalization/rw/data/__init__.py [hunk 9]
```diff
@@ -0,0 +1,13 @@
+# Copyright (c) 2023, NVIDIA CORPORATION & AFFILIATES.  All rights reserved.
+#
```
Comment (tbartley94) [inline nemo_text_processing/text_normalization/rw/data/__init__.py:2]: 2024 and Add Copyrights for Digital Umuganda (or parent org..if applicable)

File: nemo_text_processing/text_normalization/rw/__init__.py [hunk 10]
```diff
@@ -1,3 +1,17 @@
+# Copyright (c) 2023, NVIDIA CORPORATION & AFFILIATES.  All rights reserved.
+#
```
Comment (tbartley94) [inline nemo_text_processing/text_normalization/rw/__init__.py:2]: 2024 and Add Copyrights for Digital Umuganda (or parent org..if applicable)

File: nemo_text_processing/text_normalization/rw/data/cardinals/__init__.py [hunk 11]
```diff
@@ -0,0 +1,13 @@
+# Copyright (c) 2022, NVIDIA CORPORATION & AFFILIATES.  All rights reserved.
+#
```
Comment (tbartley94) [inline nemo_text_processing/text_normalization/rw/data/cardinals/__init__.py:2]: 2024 and Add Copyrights for Digital Umuganda (or parent org..if applicable)

**Comment by tbartley94** [comment]
@kurt0cougar Any update on this or are you waiting on me to do a review? (Apologies if latter, got swamped.)

File: nemo_text_processing/text_normalization/rw/taggers/cardinal.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (tbartley94) [reply nemo_text_processing/text_normalization/rw/taggers/cardinal.py:56]: If it's only found in the other language then nevermind.

**Comment by kurt0cougar** [comment]
> @kurt0cougar Any update on this or are you waiting on me to do a review? (Apologies if latter, got swamped.)
I am almost done. pytest results and Sparrowhawk tests are not generating the same results on cardinals, I am working on fixing this.

**Comment by tbartley94** [comment]
> > @kurt0cougar Any update on this or are you waiting on me to do a review? (Apologies if latter, got swamped.)
> 
> I am almost done. pytest results and Sparrowhawk tests are not generating the same results on cardinals, I am working on fixing this.

Sounds good, feel free to let us know if we can help.

**Comment by tbartley94** [comment]
@kurt0cougar Is this ready for another review pass or still working on pushes?

**Comment by kurt0cougar** [comment]
> @kurt0cougar Is this ready for another review pass or still working on pushes?

Still working on pushes, it seems like I have some conflicts to solve.

**Pull request closed without merging**